### PR TITLE
Handle notable trait popup differently

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
POC for https://github.com/rust-lang/rust/issues/91290.

So this is what I had in mind: having a `notable-traits.js` file in each crate and load it only when needed (I replicated what @jsha did for the search).

Currently, it's only on click but we can also add "mouseover", I didn't do it in this POC for simplicity. To handle the click, I simply changed the current code. When clicking on "i", we move and update the popup to display the notable trait information.

The advantage of this method is that we have element to care about and that's it and our HTML is finally valid based on the HTML spec.

You can try it out [here](https://rustdoc.crud.net/imperio/notable-trait-popup-handling/std/index.html).

Ah also, size of the rendering before/after:

```console
$ cd build/x86_64-unknown-linux-gnu/doc/std
$ du -sh .
106M	.
$ du -sh .
100M	.
```

So not much difference.

cc @jsha @Manishearth

### CODE REVIEW

**Comment by Manishearth** [comment]
I'd really like to avoid having a _second_ mega JS file that contains stuff for all the crates combined: Can we include the info on the generated file itself? Either as `data` attributes on the notable popup button, or as `display:none` HTML elements as siblings.

Yes, the current way we generate HTML is nonconformant, but the fix to that is to make it a sibling not a child.

**Comment by jsha** [comment]
> I'd really like to avoid having a second mega JS file that contains stuff for all the crates combined: Can we include the info on the generated file itself? Either as data attributes on the notable popup button, or as display:none HTML elements as siblings.

I'm in favor of the sibling approach too. To be clear: a sibling of the `<summary>` tag. My reasoning is that it's easier to inspect the HTML and reason about it when it's part of the DOM tree.

My proposed optimization on the issue was that we'd have a pile of such sibling at the end of each generated HTML page, and select the appropriate one. But that's probably over-optimizing - it's not clear how many pages (other than `Vec`) have large numbers of duplicated Notable Traits.

**Comment by GuillaumeGomez** [comment]
> I'd really like to avoid having a second mega JS file that contains stuff for all the crates combined

Sorry if I wasn't clear: each file is crate specific. It's not in the root but in the crate (contrary to the search index).

> Can we include the info on the generated file itself? Either as data attributes on the notable popup button, or as display:none HTML elements as siblings.

I'm not a big fan of the sibling approach because in the case it's in a `<summary>`, it's going to be tricky to put it into the right location (ie: not in the `<summary>`).

> Either as data attributes on the notable popup button,

As data attribute could work. It'll just be a very big one haha.

> I'm in favor of the sibling approach too. To be clear: a sibling of the <summary> tag.

The issue with this is that it might "conflict" with the `docblock` element (which is the one following the item), to be confirmed. Another issue is for the implementation in rustdoc, I think it'll make a lot of code to handle it correctly.

What about simply putting it at the end of the DOM?

> My proposed optimization on the issue was that we'd have a pile of such sibling at the end of each generated HTML page, and select the appropriate one. But that's probably over-optimizing - it's not clear how many pages (other than Vec) have large numbers of duplicated Notable Traits.

I think it's the best approach because it'd be much simpler to handle. Sibling with `<summary>` context doesn't seem like a good idea to me.

**Comment by Manishearth** [comment]
Thinking about it more, sibling of the summary tag won't work well if we have multiple of these. It's kinda unfortunate that we're putting the entire signature inside the `summary` tag but it makes some sense.

A way to still have HTML work would be to generate them all at the end and give them ids. Not great. I'm still not a fan of generating these at runtime since it moves more generation logic into JS, but I'm not sure if this is any better.

**Comment by GuillaumeGomez** [comment]
It can be pure HTML at the end of the current file too, doesn't have to be pure JS. But in any case, the display will need to be handled by JS unfortunately.

**Comment by Manishearth** [comment]
> But in any case, the display will need to be handled by JS unfortunately.

There's a crucial distinction: If it's HTML at the end of the file, JS never needs to _construct_ HTML, it simply needs to toggle some on-element CSS properties.

**Comment by GuillaumeGomez** [comment]
I'm fine with both approach. What do you think @jsha?

Just to sum it up:

 * Storing as JS allows to save a bit of space but we then have to build some HTML.
 * Storing as HTML only requires to change a CSS property to make it work (meaning much less JS).

**Comment by Manishearth** [comment]
Yep, that's an accurate summary. I'm not a fan of either approach but I have a slight preference to the HTML one from a code architecture point of view.

File: src/librustdoc/html/render/context.rs [hunk 1]
```diff
@@ -129,6 +129,9 @@ crate struct SharedContext<'tcx> {
     crate cache: Cache,
 
     crate call_locations: AllCallLocations,
+    /// The key is the notable trait text and the value is its index (we add it in the DOM so the
+    /// JS knows which notable trait to pick).
+    crate notable_traits: RefCell<FxHashMap<(String, String), usize>>,
```
Comment (camelid) [inline src/librustdoc/html/render/context.rs:134]: Are you absolutely sure this needs a RefCell?

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/context.rs:134]: For now at least it does unfortunately...

**Comment by GuillaumeGomez** [comment]
I moved the DOM directly into the item's page. So now the JS only display it (in the correct location).

File: src/librustdoc/html/render/context.rs (hunk 1 above)
Comment (camelid) [reply src/librustdoc/html/render/context.rs:134]: Ah, I didn't notice this is in the SharedContext.

**Comment by Dylan-DPC** [comment]
Closing this after a discussion with the author

**Pull request closed without merging**